CMakeLists.txt

```diff
@@ -63,8 +64,21 @@ configure_file("${CMAKE_SOURCE_DIR}/cmake/templates/license.h.in" include/licens
 
 include(EthOptions)
 configure_project(TESTS)
+set(LATEST_Z3_VERSION "4.8.12")
+if (${STRICT_Z3_VERSION})
+    find_package(Z3 ${LATEST_Z3_VERSION} EXACT)
+    if (${USE_Z3} AND NOT ${Z3_FOUND})
+      message(
+        FATAL_ERROR
+        "SMTChecker tests require Z3 ${LATEST_Z3_VERSION} for all tests to pass.\n\
+Build with -DSTRICT_Z3_VERSION=OFF if you want to use an older version. \
+You can also use -DUSE_Z3=OFF to build without Z3. In both cases use --no-smt when running tests."
+            )
+    endif()
+else()
+    find_package(Z3 4.8.0 REQUIRED)
```

Comment:
Well, we did not say we want an error with `STRICT_Z3_VERSION=OFF` when Z3 is missing. Only that we want an error when it's present but too old.

But I'd personally be fine with this. Requiring `-DUSE_Z3=OFF` when you do not have Z3 seems reasonable to me. @leonardoalt what do you think?

Reply:
Is that the default though?
I'd say it should work fine if there's no z3 available by default.

Reply:
`-DUSE_Z3` is `ON` by default so without Z3 you would get an error unless you set it to `OFF`.

Reply:
> Well, we did not say we want an error with `STRICT_Z3_VERSION=OFF` when Z3 is missing. Only that we want an error when it's present but too old.
> 
@cameel 
You're right, that way we would have two different behaviors in case of setting the `STRICT_Z3_VERSION` to off and on:
off :: if z3 is not installed we error out. If it's installed it must be ='4.8.12'
on :: if z3 is not installed it's ok. if it's installed it must be >='4.8.0'
So I implemented  it this way:
off :: if z3 is not installed we error out, If it's installed it must be ='4.8.12'
on :: if z3 is not installed we error out, if it's installed it must be >='4.8.0'

I see @leonardoalt  prefers the following:
off :: if z3 is not installed it's ok. If it's installed it must be ='4.8.12'
on :: if z3 is not installed it's ok. if it's installed it must be >='4.8.0'

I'm ok with the solution as long as we have same behavior when z3 is not installed in both cases of STRICT_Z3_VERSION is on or off.

Reply:
There's still a problem with the current version:
```cmake
if (${STRICT_Z3_VERSION})
  find_package(Z3 ${LATEST_Z3_VERSION} EXACT)
  if (${USE_Z3} AND NOT ${Z3_FOUND})
    message(
      WARNING
      "SMTChecker tests require Z3 ${LATEST_Z3_VERSION} for all tests to pass.\n\
Build with -DSTRICT_Z3_VERSION=OFF if you want to use an older version. \
You can also use -DUSE_Z3=OFF to build without Z3. In both cases use --no-smt when running tests."
          )
  endif()
else()
  find_package(Z3 4.8.0)
endif()
```

We actually have 3 cases, not 2 and I think that this code will not produce an error in cases 1.2 and 2.2 below:
1. `STRICT_Z3_VERSION=ON`
    1. Z3 not installed: OK or ERROR (I'm fine with both, @leonardoalt prefers OK)
    2. Z3 != 4.8.12: ERROR
    3. Z3 == 4.8.12: OK
2. `STRICT_Z3_VERSION=OFF`
    1. Z3 not installed: OK or ERROR (I'm fine with both, @leonardoalt prefers OK)
    2. Z3 < 4.8.0: ERROR
    3. Z3 >= 4.8.0: OK

It will give us a warning in 1.2 but will still let us continue.

Reply:
I don't follow,  I didn't call find_package with `REQUIRED` so I think both cases 1.2 and 2.2 does not error out.
Maybe you thought find_package gives error by default, but it only gives error and stop the configuration process if we call it with `REQUIRED`.